src/passes/I64ToI32Lowering.cpp

```diff
@@ -38,6 +38,48 @@ static Name makeHighName(Name n) {
 }
 
 struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
+  struct TempVar {
+    TempVar(Index idx, I64ToI32Lowering& pass) :
+      idx(idx), pass(pass), moved(false) {}
+
+    TempVar(TempVar&& other) : idx(other), pass(other.pass), moved(false) {
```

Comment:
i wonder if it's important to handle moving here? is it for correctness, or to save locals? if just to save locals, we can depend on opts to fix that up later, and this logic could be simpler.

Reply:
No, being move-only is critical for the correctness of these vars. When an expression is lowered its high bits are stored in a new variable and this variable is stored in a lookup table to be retrieved by the expression's parent. The parent can choose to reuse or free the variable, but until the parent expression is processed, the variable must not be used for anything else.

By only allowing variables to be moved and not copied, we can ensure that a variable cannot be improperly reused. Without the move semantics it would be impossible to use RAII for these variables since they would be freed and potentially reused before the parent expression could consume them.

Reply:
Got it, thanks.